### PR TITLE
Disallow applying inappropriate resource files, fix ResourceManager checking for resource

### LINKED ISSUE #177: parallel-installation CRDs installation improvement
**Description**

At the moment we have a job in the cluster-essentials component, which is applying CRDs. Move this logic directly to the library and implement retries.

AC:
- CRDs installed via k83 client
- retries implemented
- if pod-preset is still present in the cluster-essentials install only this sub-chart

**Reasons**

Having an easy and stable way of installing CRDs

### PR DESCRIPTION
**Pull request by mjakobczyk**
**Description**

There have been two issues detected. The first one is that asking k8s API server for resource should not end when error different than `not found` appears.
The second one is connected with security. There should be no possibility to allow parsing resources different than CRDs or Namespaces in their corresponding directories. To achieve this goal the logic from `ResourceApplier` has been extracted to the new entity called `ResourceParser` which gives us a possibility to retrieve file metadata info before applying them straight to the k8s cluster. As a result, `PreInstaller` now does not allow resources that resource kind differ from the one passed in input (`CustomResourceDefinition` or `Namespace`  for now). Appropriate test cases have been added to cover these situation.

Changes proposed in this pull request:

- remove initial `ResourceManager#getResource` call
- move resource not found logic to the retry loop
- return nil when first encounter not found error
- introduce `ResourceParser`
- disallow resource application that type from the one to corresponding method (`InstallCRDs` omit resources whose kind does not match `CustomResourceDefinition` and the same for `CreateNamespaces` - only `Namespace` kind is allowed)

**Related issue(s)**

Fixes #177

### CODE REVIEW

File: parallel-install/pkg/preinstaller/resourceparser.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+		return obj, err
+	}
+
+	return c.parseResourceFrom(string(manifest))
+}
+
+func (c *GenericResourceParser) parseResourceFrom(manifest string) (*unstructured.Unstructured, error) {
+	var _, _, err = c.decoder.Decode([]byte(manifest), nil, nil)
+	if err != nil {
+		return nil, errors.New(fmt.Sprintf("Could not decode the resource file due to the following error: %s.", err.Error()))
```
Comment (clebs) [inline parallel-install/pkg/preinstaller/resourceparser.go:46]: Please use `errors.Wrap()` when wrapping an error into another.

File: parallel-install/pkg/preinstaller/resourceparser.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+	resource := &unstructured.Unstructured{
+		Object: middleware,
+	}
+	return resource, nil
+}
+
+func initializeDecoder() runtime.Decoder {
+	sch := runtime.NewScheme()
+	_ = clientgoscheme.AddToScheme(sch)
+	_ = apiextv1.AddToScheme(sch)
```
Comment (clebs) [inline parallel-install/pkg/preinstaller/resourceparser.go:90]: Please do not ignore errors.

File: parallel-install/pkg/preinstaller/mocks/resourceparsermock.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
+package mocks
```
Comment (clebs) [inline parallel-install/pkg/preinstaller/mocks/resourceparsermock.go:1]: For autogenerated and more flexible mocks, please check `mockery`.
You can find a full usage here:
https://github.com/kyma-project/cli/blob/bad193c82b5fa538f69335789e64bb536beb2200/internal/kube/kube.go#L15-L16

and how to use it:
https://github.com/kyma-project/cli/blob/fa37f116219247daae3487abad5499cc6fa939ec/cmd/kyma/create/system/system_test.go#L67-L68 

Writing your own mocks is a lot of unnecessary work and it adds more maintenance effort as the code evolves.
Generated mocks take all that away.

File: parallel-install/pkg/preinstaller/preinstaller.go [hunk 4]
```diff
@@ -180,14 +182,28 @@ func (i *PreInstaller) apply(resources []resourceInfoResult) (o Output, err erro
 			path:      resource.path,
 		}
 
+		parsedResource, err := i.parser.ParseUnstructuredResourceFrom(file.path)
+		if err != nil {
+			i.cfg.Log.Warn(fmt.Sprintf("Error occurred when processing resource %s of component %s : %s", resource.fileName, resource.component, err.Error()))
```
Comment (clebs) [inline parallel-install/pkg/preinstaller/preinstaller.go:187]:
```suggestion
			i.cfg.Log.Warnf("Error occurred when processing resource %s of component %s : %s", resource.fileName, resource.component, err.Error())
```

File: parallel-install/pkg/preinstaller/preinstaller.go [hunk 5]
```diff
@@ -180,14 +182,28 @@ func (i *PreInstaller) apply(resources []resourceInfoResult) (o Output, err erro
...
+			o.NotInstalled = append(o.NotInstalled, file)
+			continue
+		}
+
+		if parsedResource.GetKind() != resource.resourceType {
+			i.cfg.Log.Warn(fmt.Sprintf("Resource type does not match for resource %s of component %s : got %s but expected %s", resource.fileName, resource.component, parsedResource.GroupVersionKind().Kind, resource.resourceType))
```
Comment (clebs) [inline parallel-install/pkg/preinstaller/preinstaller.go:193]:
```suggestion
			i.cfg.Log.Warnf("Resource type does not match for resource %s of component %s : got %s but expected %s", resource.fileName, resource.component, parsedResource.GroupVersionKind().Kind, resource.resourceType)
```

File: parallel-install/pkg/preinstaller/preinstaller.go [hunk 6]
```diff
@@ -180,14 +182,28 @@ func (i *PreInstaller) apply(resources []resourceInfoResult) (o Output, err erro
...
+			o.NotInstalled = append(o.NotInstalled, file)
+			continue
+		}
+
 		i.cfg.Log.Info(fmt.Sprintf("Processing %s file: %s of component: %s", resource.resourceType, resource.fileName, resource.component))
```
Comment (clebs) [inline parallel-install/pkg/preinstaller/preinstaller.go:198]:
```suggestion
		i.cfg.Log.Infof("Processing %s file: %s of component: %s", resource.resourceType, resource.fileName, resource.component)
```

File: parallel-install/pkg/preinstaller/preinstaller.go [hunk 7]
```diff
@@ -180,14 +182,28 @@ func (i *PreInstaller) apply(resources []resourceInfoResult) (o Output, err erro
...
-		err = i.applier.Apply(file.path)
+		err = i.applier.Apply(parsedResource)
 		if err != nil {
 			i.cfg.Log.Warn(fmt.Sprintf("Error occurred when processing file %s of component %s : %s", resource.fileName, resource.component, err.Error()))
```
Comment (clebs) [inline parallel-install/pkg/preinstaller/preinstaller.go:201]:
```suggestion
			i.cfg.Log.Warnf("Error occurred when processing file %s of component %s : %s", resource.fileName, resource.component, err.Error())
```

File: parallel-install/pkg/preinstaller/resourceparser.go [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+type ResourceParser interface {
+	// ParseUnstructuredResourceFrom given path and return it.
+	ParseUnstructuredResourceFrom(path string) (*unstructured.Unstructured, error)
+}
+
+// GenericResourceParser is a default implementation of ResourceParser.
+type GenericResourceParser struct{}
+
+// NewGenericResourceParser returns a new instance of GenericResourceParser.
+func NewGenericResourceParser() *GenericResourceParser {
```
Comment (clebs) [inline parallel-install/pkg/preinstaller/resourceparser.go:25]: If a constructor is not doing anything on the struct, it is not needed. One can just instantiate the struct:
`grp := &preinstaller.GenericResourceParser{}`

**Review by clebs — CHANGES_REQUESTED** [review]

File: parallel-install/pkg/preinstaller/preinstaller.go [hunk 9]
```diff
@@ -180,14 +182,28 @@ func (i *PreInstaller) apply(resources []resourceInfoResult) (o Output, err erro
 			path:      resource.path,
 		}
 
-		i.cfg.Log.Info(fmt.Sprintf("Processing %s file: %s of component: %s", resource.resourceType, resource.fileName, resource.component))
-		err = i.applier.Apply(file.path)
+		parsedResource, err := i.parser.ParseUnstructuredResourceFrom(file.path)
```
Comment (Tomasz-Smelcerz-SAP) [inline parallel-install/pkg/preinstaller/preinstaller.go:185]:
```suggestion
		parsedResource, err := i.parser.ParseFile(file.path)
```

File: parallel-install/pkg/preinstaller/resourcemanager.go [hunk 10]
```diff
@@ -53,34 +55,27 @@ func (c *DefaultResourceManager) CreateResource(resource *unstructured.Unstructu
...
+			if apierrors.IsNotFound(err) {
+				c.log.Infof("Resource %s was not found.", resourceName)
+				return nil
+			} else {
+				c.log.Errorf("Error occurred during resource get: %s", err.Error())
+				return err
+			}
 		}
 
 		return err
```
Comment (Tomasz-Smelcerz-SAP) [inline parallel-install/pkg/preinstaller/resourcemanager.go:74]: The code would be a bit more readable if a `nil` is returned here.

**Review by Tomasz-Smelcerz-SAP — APPROVED** [review]

**Review by clebs — APPROVED** [review]

**Pull request merged**